compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/Fir2IrConverter.kt

```diff
@@ -94,14 +94,31 @@ class Fir2IrConverter(
         return irClass
     }
 
+    // Sort declarations so that all non-synthetic declarations are before synthetic ones.
+    // This is needed because converting synthtic fields for implementation delegation needs to know
+    // existing declarations in the class to avoid adding redundant delegated members.
+    private fun putSyntheticLast(declarations: List<FirDeclaration>) : List<FirDeclaration> {
+        val result = mutableListOf<FirDeclaration>()
+        declarations.forEach {
+            if (it.origin != FirDeclarationOrigin.Synthetic) {
+                result.add(it)
+            }
+        }
+        declarations.forEach {
+            if (it.origin == FirDeclarationOrigin.Synthetic) {
+                result.add(it)
+            }
+        }
+        return result
```

Comment:
These are readable, so I think it would be okay. I have a couple thoughts, though.

1. groupBy: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/group-by.html

This seems to me like grouping synthetic v.s. non-synthetic members, something like:
```
declarations.groupBy { it.origin == FirDeclarationOrigin.Synthetic }.values.flatten()
```
assuming `false` key is placed ahead (otherwise, well, flipping the condition :) ).

2. materialization

I'm not sure there will be a huge difference or a class with too many declarations, but I wonder we could avoid creating another materialized list here. There are only two uses (above and below), which are simply iteration. So, we could return either `Iterable` or `Sequence` instead?

Reply:
done